### ISSUE TITLE
why retinanet outperforms maskrcnn for lung lesion detection?

### ISSUE THREAD

**Issue by gathierry** [MIC-DKFZ/medicaldetectiontoolkit#25]
Hi,
I've read your paper and I'm wondering why RetinaNet works better than MaskRCNN for lung lesion detection with 2D or 2Dc inputs. I think in most cases, a two-stage detector should be better than one-stage with the same backbone. Have you ever investigate this? Or did I miss any details?
Thanks

**Comment by pfjaeger** [comment]
Hi,

Thanks for reaching out. Why would a 2 stage detector be better in your opinion ?

Sent from my iPhone

On 11. Mar 2019, at 10:16, Shiyu <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi,
I've read your paper and I'm wondering why RetinaNet works better than MaskRCNN for lung lesion detection with 2D or 2Dc inputs. I think in most cases, a two-stage detector should be better than one-stage with the same backbone. Have you ever investigate this? Or did I miss any details?
Thanks

-

**Comment by gathierry** [comment]
For common object detection like COCO, MaskRCNN works better than RetinaNet. But their configurations are not exactly the same, I have to admit. One has additional instance labels and the other one uses focal loss.
In my opinion, a one-stage model is more or less like an RPN. The second stage can eliminate false positives and fine-tune the regression boxes.
Do you think this explanation makes sense for lesion detection?

**Comment by pfjaeger** [comment]
The resampling operation (RoiPool / RoiAlign) was initiated to enforce scale-invariance of box-proposals. This only makes sense in natural images, so basically you want to re-scale a big car and a small car to the same size to facilitate the classification task for the classifier head. Here, it makes sense to get rid of the scale information, because it is an artifact caused only by varying distances between camera and objects which contains no semantic information. In medical images we assume that the distance between camera and object is the same over the entire image, thus the scale-information contains semantically relevant information such as size of a tumor. Intuitively, you would not want to cancel this out in your model. 

There is another reason, why the second stage could rather hurt performance than boost it: As there is no reason for the re-sampling operation, there is also no real benefit from the second stage operations, because no new information is gained/processed. In an ideal decision process, you would want to gather all information possible and then take one informed decision (this is what happens in a one stage detector), as opposed to a sequential decision process, where information gets discarded along the way possibly throwing away relevant bits, before taking the ultimate decision (two stage models).

So in fact, there are two reasons why one stage models should intuitively be superior to two stage models on medical images.

Do you agree?

**Comment by gathierry** [comment]
Thanks, I think this makes good sense. The scale information a priori can be used by the model. But does this mean there's risk of overfitting if the scale distribution of training set is different from the test set? I think this is possible for medical imaging data.

**Comment by pfjaeger** [comment]
yes good point. If your test data contains lesion sizes not apparent in the training data, you might be better off using 2 stage detectors, because here the re-sampling helps you to overcome the domain shift. I would, however, call this a rather special case, as arguably most of machine learning apps (that are not specifically dealing with domain shifts) assume equal distributions in training and test data.

**Comment by gathierry** [comment]
Yes, I agree. Different distributions between training and test set is quite challenging and only a two-stage detector could be still no enough.
Thanks again for the great explanation, and this nice repo.

**Comment by pfjaeger** [comment]
glad I could help. thanks for your feedback.
